pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -201,36 +213,50 @@ async def maybe_nack(ack_id: str) -> None:
                 log.warning(
                     'Nack request failed, better luck next batch', exc_info=e)
                 metrics_client.increment('pubsub.nacker.batch.failed')
+                metrics.BATCH_STATUS.labels(
+                    component='nacker', outcome='failed').inc()
 
                 continue
 
             metrics_client.histogram('pubsub.nacker.batch', len(ack_ids))
+            metrics.BATCH_STATUS.labels(component='nacker',
+                                        outcome='succeeded').inc()
+            metrics.MESSAGES_PROCESSED.labels(component='nacker').inc(
+                len(ack_ids))
 
             ack_ids = []
 
     async def _execute_callback(message: SubscriberMessage,
                                 callback: ApplicationHandler,
                                 ack_queue: 'asyncio.Queue[str]',
                                 nack_queue: 'Optional[asyncio.Queue[str]]',
+                                insertion_time: float,
                                 metrics_client: MetricsAgent
                                 ) -> None:
         try:
             start = time.perf_counter()
-            await callback(message)
-            await ack_queue.put(message.ack_id)
-            metrics_client.increment('pubsub.consumer.succeeded')
+            metrics.CONSUME_LATENCY.labels(phase='queueing').observe(
+                start - insertion_time)
+            with metrics.CONSUME_LATENCY.labels(phase='runtime').time():
+                await callback(message)
+                await ack_queue.put(message.ack_id)
             metrics_client.histogram('pubsub.consumer.latency.runtime',
                                      time.perf_counter() - start)
+            metrics_client.increment('pubsub.consumer.succeeded')
+            metrics.CONSUME.labels(outcome='succeeded').inc()
+
         except asyncio.CancelledError:
             if nack_queue:
                 await nack_queue.put(message.ack_id)
             log.warning('Application callback was cancelled')
             metrics_client.increment('pubsub.consumer.cancelled')
+            metrics.CONSUME.labels(outcome='cancelled').inc()
         except Exception:
             if nack_queue:
                 await nack_queue.put(message.ack_id)
             log.exception('Application callback raised an exception')
             metrics_client.increment('pubsub.consumer.failed')
+            metrics.CONSUME.labels(outcome='failed').inc()
```

Comment:
I wonder, is it possible for applications to inject their own labels to extend these metrics? It would be wonderful to solely rely on this library for handling all of the metric capturing, but at the same time we often include additional information with latency measurements and such. For example in realtime-egress we break down the number of failures by `stage`, `domain` and `entity_type`, and in `realtime-nlp` we break down runtime latency by whether or not the model emitted any `events` (not to mention which model emitted them)

Reply:
Adding labels at runtime isn't really something prometheus supports; metrics need to have all their metrics declared upfront. I think the "default" way to do what you're describing with prometheus is to have your app layer just create _new_ metrics as needed.

Reply:
Yeah that's what I was thinking, which is already what we are currently doing and thus are essentially duplicating some of our metric capturing. I was just wondering if we could clean up some of our application code (the metric capturing portion) in apps which make use of this library.